src/core/qgsmaplayer.cpp

```diff
@@ -98,6 +98,11 @@ QgsMapLayer::QgsMapLayer( QgsMapLayerType type,
 
 QgsMapLayer::~QgsMapLayer()
 {
+  if ( QgsProject::instance()->pathResolver().writePath( mDataSource ).startsWith( "attachment:" ) )
+  {
+    QgsProject::instance()->removeAttachedFile( mDataSource );
+  }
```

Comment:
Is `project()` a feasible replacement for the singleton call here?

Reply:
:+1: